pkg/apis/apps/v1alpha1/apimanager_types.go

```diff
@@ -513,59 +504,36 @@ func (apimanager *APIManager) setSystemSpecDefaults() (bool, error) {
 	return changed, nil
 }
 
-func (apimanager *APIManager) setSystemFileStorageSpecDefaults() (bool, error) {
+func (apimanager *APIManager) setSystemFileStorageSpecDefaults() bool {
 	changed := false
 	systemSpec := apimanager.Spec.System
 
-	defaultFileStorageSpec := &SystemFileStorageSpec{
-		PVC: &SystemPVCSpec{
-			StorageClassName: nil,
-		},
-	}
-
 	if systemSpec.FileStorageSpec == nil {
-		systemSpec.FileStorageSpec = defaultFileStorageSpec
+		systemSpec.FileStorageSpec = &SystemFileStorageSpec{}
 		changed = true
-		return changed, nil
 	}
 
-	if systemSpec.FileStorageSpec.PVC != nil && systemSpec.FileStorageSpec.S3 != nil {
-		return changed, fmt.Errorf("Only one FileStorage can be chosen at the same time")
-	}
-
-	if systemSpec.FileStorageSpec.PVC == nil && systemSpec.FileStorageSpec.S3 == nil {
-		systemSpec.FileStorageSpec.PVC = defaultFileStorageSpec.PVC
+	// No default values for S3
+	// PVC
+	if systemSpec.FileStorageSpec.S3 == nil && systemSpec.FileStorageSpec.PVC == nil {
+		systemSpec.FileStorageSpec.PVC = &SystemPVCSpec{}
 		changed = true
+		// StorageClassName default value is nil
 	}
 
-	return changed, nil
+	return changed
 }
 
-func (apimanager *APIManager) setSystemDatabaseSpecDefaults() (bool, error) {
+func (apimanager *APIManager) setSystemDatabaseSpecDefaults() bool {
 	changed := false
 	systemSpec := apimanager.Spec.System
 
-	defaultDatabaseSpec := &SystemDatabaseSpec{
-		MySQL: &SystemMySQLSpec{
-			Image: nil,
-		},
-	}
-
 	if systemSpec.DatabaseSpec == nil {
-		systemSpec.DatabaseSpec = defaultDatabaseSpec
+		systemSpec.DatabaseSpec = &SystemDatabaseSpec{}
 		changed = true
-		return changed, nil
 	}
 
-	if systemSpec.DatabaseSpec.MySQL != nil && systemSpec.DatabaseSpec.PostgreSQL != nil {
```

Comment:
What's the motivation of moving CR validation to reconciliation?
To me conceptually it makes more sense to validate the schema before applying the reconciliation logic. I think ideally the reconciliation should take a valid CR / Internal object

Reply:
Syntactic validation and at some extent semantic validation is done by CRD openapi validation. Most of the higher level semantic validation is done by option builders, in the reconciliation loop. Hence, moving validation to the reconciliation loop in a distributed way, i.e. each components validates each own part.

I also like having a single entrypoint at the beginning of the reconciliation loop to do all validation checks for high level semantic issues. 

Definitely, validation at a method meant to set defaults is not conceptually correct.

Reply:
I don't like the idea of validating on the reconciliation loop. I'd prefer knowing that when the reconciliation logic is applied we start from valid contents.

I thought of having a validate method separate from the setDefaults too @eguzki in apimanager_types.go and I'd like to go that way but I found the following dilemma, which is why I left validation on the defaults method:

What/when is considered a CR "valid"? Should the validation be performed before or after setting the defaults?

You can have an "invalid" CR that "becomes valid" after defaults are set for it. For example, the replicas fields are optional but we set them as we need a value for them defined. Same happens for the selected storage type.

If you "validate" before setting the defaults you might detect that the CR is invalid but it is not really invalid, it's just that the defaults have not been set for it yet (replicas or storage case explained before).
If you "validate" after setting the defaults then you might try to set defaults from fields that are not initialized and that you expect them to be (a really invalid object). Shouldn't defaults be applied only when we know that we have a valid object?

What do you think?

Reply:
As I see:
Validation should be performed **before** applying defaults. Applying defaults always **adds** info, never update or delete.

> You can have an "invalid" CR that "becomes valid" after defaults are set for it

I disagree. CR without defaults is also valid, as defaults are always optional fields. As I see, invalid state is when, for instance, required attribute does not exist, or when two excluding fields exist (like pvc and s3). As I see, default concept lives in the code. If user does not set replicas, the object still have the `replica` field with some value set by the code. That is being optional with a default value. Hence, replica will always exist internally, even when the user does not set explicitly. So, it will be valid even when replica is not set.

As I said, the operator should validate before applying defaults. It could happen that after validating and after setting defaults the CR could be invalid. But in that case, this is a bug we should fix.

I do not see any good reason to apply defaults to the CR. We could "serialize" the internal CR object in status if we want to report that info to the customer.

Reply:
> I do not see any good reason to apply defaults to the CR. We could "serialize" the internal CR object in status if we want to report that info to the customer.

I think that's not what status should be used for. Status is the "actual" status whereas "spec" is the desired status.
I don't think it's a bad idea to apply defaults to the CR. As said in the API conventions https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#defaulting incorportaing the default values into the Spec ensures that the Spec depicts the full desired state so that it is easier for the system to determine how to achieve the state, and for the user to know what to anticipate.